src/PathSimulator.cpp

```diff
@@ -0,0 +1,395 @@
+#include "PathSimulator.hpp"
+
+#include <iterator>
+#include <utility>
+
+PathSimulator::SimulationPath::SimulationPath(std::size_t nleaves, PathSimulator::SimulationPath::ComponentsList components, const qc::QuantumComputation* qc, bool assumeCorrectOrder):
+    components(std::move(components)), nleaves(nleaves), qc(qc) {
+    steps.reserve(nleaves);
+    // create empty vector of steps
+    for (std::size_t id = 0; id < nleaves; ++id) {
+        steps.emplace_back(id, std::vector{id});
+    }
+
+    for (auto& [leftID, rightID]: this->components) {
+        if (leftID >= steps.size())
+            throw std::runtime_error("Left simulation path index out of range.");
+        if (rightID >= steps.size())
+            throw std::runtime_error("Right simulation path index out of range");
+
+        auto& leftStep  = steps.at(leftID);
+        auto& rightStep = steps.at(rightID);
+
+        const std::size_t resultID = steps.size();
+        leftStep.parent            = resultID;
+        rightStep.parent           = resultID;
+
+        std::vector<std::size_t> operations{};
+        if (!assumeCorrectOrder) {
+            // move all operations from the left and the right to this result
+            auto& leftOps  = leftStep.operations;
+            auto& rightOps = rightStep.operations;
+
+            // consider each operation from the left and check whether any operation from the right must precede it
+            bool leftIsActuallyLeft = true;
+            //        std::cout << "Checking whether ID " << leftID << " is actually left of ID " << rightID << std::endl;
+            for (const auto& leftOpID: leftOps) {
+                // if the initial state is included in the left, this has to be left
+                if (leftOpID == 0) {
+                    //                std::cout << "Left contains initial state, so it has to be left" << std::endl;
+                    break;
+                }
+                const auto& leftOp = qc->at(leftOpID - 1);
+
+                for (const auto& rightOpID: rightOps) {
+                    // if the initial state is included in the right, the right side should be left
+                    if (rightOpID == 0) {
+                        //                    std::cout << "Right contains initial state, so it has to be left" << std::endl;
+                        leftIsActuallyLeft = false;
+                        break;
+                    }
+
+                    // in case the left ID is smaller than the right ID, everything is in order
+                    if (leftOpID < rightOpID) {
+                        //                    std::cout << "Operation IDs " << leftOpID << " and " << rightOpID << " are in order." << std::endl;
+                        continue;
```

Comment:
This line should be easy to cover in the false order test by including one tuple that is in the correct order